### PR TITLE
[ADD] [16.0] sale_block_no_stock

### PR DESCRIPTION
**Pull request by Shide**
This module extends the functionality of Sales to support blocking sales and to allow you to ensure you are capable to send product you have in stock.

When a Sale Order is going to be confirmed, it will be checked if the quantity demanded exceeds that of the field selected in the configuration, and in this case an error will be thrown indicating which lines have an error, indicating what is the maximum quantity that can be ordered.

If the user who is confirming an order has a group that is allowed, it will not block the order confirmation.

https://www.loom.com/share/f737e8c5845748b7a0c6e157a4a7da0d?sid=df2a827c-4959-4163-83a8-a4c31c03d306

MT-6090 @moduon @rafaelbn @yajo @EmilioPascual @fcvalgar @Gelojr please review if you want

### CODE REVIEW

**Review by fcvalgar — APPROVED** [review]
Functional review, 

LGTM thank you @Shide 

image


image

image

File: sale_block_no_stock/readme/USAGE.md
Comment (EmilioPascual) [inline sale_block_no_stock/readme/USAGE.md:1]: This documentation actually explains the configuration parameters, not the use of the module.

**Review by EmilioPascual — APPROVED** [review]
@Shide good job.
Just a small comment but i approve anyway

File: sale_block_no_stock/readme/USAGE.md
Comment (Shide) [reply sale_block_no_stock/readme/USAGE.md:1]: Doc updated

**Comment by rafaelbn** [comment]
I've review it, we need to wait for some more use cases + QA. Thank you for waiting!

File: sale_block_no_stock/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def action_confirm(self):
+        qty_exceptions = []
+        for record in self:
+            if any(
+                self.env.user in group.users
+                for group in record.company_id.sale_line_block_allowed_groups
+            ):
```
Comment (yajo) [inline sale_block_no_stock/models/sale_order.py]: suggestion: perf improvement

```suggestion
            if self.env.user in record.company_id.sale_line_block_allowed_groups.users:
```

File: sale_block_no_stock/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    sale_line_field_block = fields.Many2one(
+        comodel_name="ir.model.fields",
+        string="Field to compare quantities on Sale Order Lines",
```
Comment (yajo) [inline sale_block_no_stock/models/res_company.py:12]: suggestion: I think specifying the type of quantity or check will help understand how this works.

```suggestion
        string="Field to compare available quantities on Sale Order Lines",
```

File: sale_block_no_stock/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                    "free_qty_today",
+                    "qty_available_today",
+                    "virtual_available_at_date",
+                ),
+            ),
+        ],
+    )
+    sale_line_block_allowed_groups = fields.Many2many(
+        comodel_name="res.groups",
+        string="Allowed Groups to bypass the block",
```
Comment (yajo) [inline sale_block_no_stock/models/res_company.py:30]:
```suggestion
        string="Allowed Groups to bypass the availability block",
```

File: sale_block_no_stock/views/res_config_settings_views.xml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                            class="fa fa-lg fa-building-o"
+                            title="Values set here are company-specific."
+                            groups="base.group_multi_company"
+                        />
+						<div class="text-muted">
+							Avoid selling more product than you have in stock or planned
+						</div>
+						<div class="content-group">
+							<div class="mt-2">
+                                <label
```
Comment (yajo) [inline sale_block_no_stock/views/res_config_settings_views.xml:22]: It seems indentation went wild here!

File: sale_block_no_stock/views/res_config_settings_views.xml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                                    name="sale_line_block_allowed_groups"
+                                    widget="many2many_tags"
+                                    options="{'no_create': True, 'no_open': True}"
+                                    class="w-100"
+                                />
+							</div>
+						</div>
+					</div>
+				</div>
+            </xpath>
```
Comment (yajo) [inline sale_block_no_stock/views/res_config_settings_views.xml:49]: Here too

**Review by yajo — APPROVED** [review]
I found some minor points to improve, but the module is amazing and nice.

File: sale_block_no_stock/models/sale_order.py (hunk 1 above)
Comment (Shide) [reply sale_block_no_stock/models/sale_order.py]: Doesn't work like this. I've tryied this solution before

File: sale_block_no_stock/models/res_company.py (hunk 2 above)
Comment (Shide) [reply sale_block_no_stock/models/res_company.py:12]: Depends on the field you choose. Available means an specific field on the line. This way leaves the dependency of the meaning with another field.

File: sale_block_no_stock/models/res_company.py (hunk 3 above)
Comment (Shide) [reply sale_block_no_stock/models/res_company.py:30]: Same as above.

File: sale_block_no_stock/views/res_config_settings_views.xml (hunk 4 above)
Comment (Shide) [reply sale_block_no_stock/views/res_config_settings_views.xml:22]: Why? Only want to write on the right_pane. And pre-commit do its things

File: sale_block_no_stock/views/res_config_settings_views.xml (hunk 5 above)
Comment (Shide) [reply sale_block_no_stock/views/res_config_settings_views.xml:49]: It's normal because I'm only putting thinks onto right_pane

**Comment by Shide** [comment]
Total facelift @rafaelbn

File: sale_block_no_stock/wizard/sale_order_block_wizard.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        self.confirmation_allowed = False
+        for record in self:
+            if not record.sale_line_block_ids:
+                continue
+            # All lines must belong to the same company, so choose first one
+            company = record.sale_line_block_ids[0].order_id.company_id
+            record.confirmation_allowed = any(
+                self.env.user in group.users
+                for group in company.sale_line_block_allowed_groups
+            )
```
Comment (yajo) [inline sale_block_no_stock/wizard/sale_order_block_wizard.py]:
```suggestion
        self.confirmation_allowed = self.env.user in self.sale_line_block_ids.company_id.sale_line_block_allowed_groups.users
```

File: sale_block_no_stock/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def action_confirm(self):
+        if self.env.context.get("skip_block_no_stock_check"):
+            return super().action_confirm()
```
Comment (yajo) [inline sale_block_no_stock/models/sale_order.py]: issue(security): Unauthorized users shouldn't be able to forge an RPC request that includes this context to bypass the blocking. Please check user permissions before taking the context into account.

File: sale_block_no_stock/wizard/sale_order_block_wizard.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+    def confirm(self):
+        if not all(self.mapped("confirmation_allowed")):
+            raise exceptions.UserError(
+                _("You are not allowed to confirm these orders.")
+            )
+        orders = self.mapped("sale_line_block_ids.order_id")
+        orders.message_post(
+            body=_("Order confirmed with errors by %s.", self.env.user.name)
```
Comment (yajo) [inline sale_block_no_stock/wizard/sale_order_block_wizard.py]: issue: shouldn't you indicate somehow that the message posted is an internal note and not a public message? I don't think customers like to get an email with this text...

**Review by yajo — CHANGES_REQUESTED** [review]
It looks much better now.

Please @fcvalgar update your functional review; this has changed substantially.

Some issues detected.

Thanks!

File: sale_block_no_stock/wizard/sale_order_block_wizard.py (hunk 8 above)
Comment (Shide) [reply sale_block_no_stock/wizard/sale_order_block_wizard.py]: Default subtype_id is notifications, but addressed.

**Comment by rafaelbn** [comment]
@Shide we should skip services product 😄

**Comment by Shide** [comment]
I'm on it

**Comment by Shide** [comment]
@rafaelbn done and updated readme and help in functions

**Review by EmilioPascual — APPROVED** [review]

**Comment by Shide** [comment]
@rafaelbn Only appears one button: If lines has packagings, show adjust packaging, if not, show adjust uom

**Comment by Shide** [comment]
Fixed sudo's to read fields and actions

**Comment by fcvalgar** [comment]
> Fixed sudo's to read fields and actions

Can you help me with runboat? Not available

**Comment by fcvalgar** [comment]
> /ocabot rebase

@rafaelbn can you rebase?

**Review by yajo — APPROVED** [review]

**Comment by yajo** [comment]
@fcvalgar I don't know why runboat isn't linked here, but at least you can find the builds in https://runboat.odoo-community.org/webui/builds.html?repo=oca/sale-workflow&target_branch=16.0 searching for this PR number

**Review by rafaelbn — APPROVED** [review]
Tested, alpha. Funcional review OK. Improvement could be accepted.

**Pull request merged**